Automattic/wp-calypso
# Comments: Track actions progress in state

With the new Comments List approach coming up in #21155, we noticed that on bulk actions, there are race conditions preventing to fetch the correct list and count of the updated page.
This happens because we don't have a bulk endpoint, but we loop over the selected comments list and fire one update request per comment, and therefore we have no simple way to know when all of them are completed.

This PR introduces a new `ui.comments.progresses` state that keeps track of any change status and permanent delete actions (bulk, but also single), provided they have the `progressId` and `progressTotal` options set in the optional `refreshCommentListQuery` parameter.

The idea is that, when `COMMENTS_CHANGE_STATUS` or `COMMENTS_DELETE` action is dispatched, we create a new `uuid`-keyed object in the new state, with the following properties:
- `count`: is the number of times this particular progress has been updated (initialized to `0`).
- `total`: total number of times this particular progress needs to be updated to be considered complete (1 for single actions; `selectedComments.length` for bulk).
- `status`: the action performed (the new status for `COMMENTS_CHANGE_STATUS`; `'delete'` for `COMMENTS_DELETE`).
- `failed`: if the progress fails somehow (initialized to `false`).

Once one of these two actions hits their `onSuccess` handler in the data layer, the brand new `COMMENTS_PROGRESS_UPDATE` is dispatched, increasing the progress `count` by 1.

Every time any of those three actions (bearing a `progressId`) enter the `progresses` reducer, it automatically clears the state of the completed (`count === total`) progresses.
In other words, a completed progress is only tracked until the next action that affects progresses is dispatched.

The idea is to listen to the `getCommentsProgresses` selector in `CommentList`, and every time a progress is completed we fire a request for a fresh list and count to keep everything in sync.

## Future possibilities

By tracking the progress, we can actually show a progress bar for any actions.
It might seem unnecessary: all in all we're not tracking the progress of large files uploads.
Though, update actions on comments are surprisingly slow.
For example, getting a list of 20 comments takes, on my .com test site, roughly 300ms.
But posting a change status on a single comment takes around 1.5s.
Just think how much slower a shared-hosted JP site could be.

But even remaining on .com: 1.5s by itself might not be that slow to need a progress bar.
Though, by bulk updating, we'd almost end up multiplying that time for the number of affected comments (not exactly true, but still we're talking _multiple_ seconds).

## Testing instructions

Try dispatching something like this in the console:

```es6
{
  type: 'COMMENTS_CHANGE_STATUS',
  siteId: SITE_ID,
  postId: POST_ID,
  commentId: COMMENT_ID,
  status: STATUS,
  refreshCommentListQuery: {
    progressId: '91d16ec5-b529-455c-84ad-5f57f3f0a7c7',
    progressTotal: 1
  },
}
```

In the Redux dev tools, make sure it is followed up by a `COMMENTS_PROGRESS_UPDATE` action.

Now try again, but using an obviously wrong COMMENT_ID.
The following `COMMENTS_PROGRESS_UPDATE` should have: `{ options: { failed: true } }`.

---

🎉 Props to @rodrigoi for the idea! 🎉
